R/fit_baselines_one_location.R

```diff
@@ -0,0 +1,158 @@
+    cli::cli_abort("The provided {.arg temporal_resolution} does not match that of the provided {.arg target_ts}")
+  }
+
+  # figure out horizons to forecast
+  reference_date <- lubridate::ymd(reference_date) # date to which horizons are relative
+  last_data_date <- max(target_ts$time_index)
+  actual_target_dates <- reference_date + ts_temp_res * horizons
+  effective_horizons <- as.integer(actual_target_dates - last_data_date) / ts_temp_res
+  horizons_to_forecast <- 1:max(effective_horizons)
+  h_adjustments <- min(effective_horizons) - 1
+
+  if (h_adjustments > 0) {
+    cli::cli_warn(
+      "forecasts requested for a time index beyond the provided {.arg target_ts},
+        replacing missing target data with {.val {h_adjustments}} target observations"
+    )
+    target_ts <- target_ts |>
+      dplyr::group_by(dplyr::across(c("location"))) |>
+      tidyr::complete(
+        time_index = unique(c(target_ts$time_index, last_data_date + h_adjustments * ts_temp_res)),
+        fill = list(observation = target_ts$observation[target_ts$time_index == last_data_date]),
+      )
+
+    last_data_date <- max(target_ts$time_index)
+    effective_horizons <- as.integer(actual_target_dates - last_data_date) / ts_temp_res
+    horizons_to_forecast <- 1:max(effective_horizons)
+    h_adjustments <- min(effective_horizons) - 1
+  }
+
+  # get predictions for all model_variations
+  predictions <- purrr::pmap_dfr( #tibble, each 1x1 row contains predictions for 1 model
+    model_variations,
+    get_baseline_predictions,
+    target_ts = target_ts,
+    effective_horizons = horizons_to_forecast,
+    origin = ifelse(temporal_resolution == "weekly", "obs", "median"),
+    n_sim = 100000,
+    quantile_levels = quantile_levels,
+    n_samples = n_samples,
+    round_predictions = round_predictions,
+    seed = seed
+  )
+
+  # extract forecasts
+  extracted_outputs <-
+    dplyr::bind_cols(model_variations, predictions) |>
+    tidyr::unnest(cols = "forecasts") |>
+    dplyr::mutate(
+      location = unique(target_ts$location),
+      horizon = .data[["horizon"]],
+      target_end_date = last_data_date + ts_temp_res * .data[["horizon"]],
+      .before = "horizon"
+    )
+
+  # h_adjustments = 0 (min(actual_target_dates) occurs period after last observed value)
+
+  if (h_adjustments == 0) {
+    model_outputs <- extracted_outputs |>
+      dplyr::mutate(
+        reference_date = reference_date,
+        horizon = as.numeric((.data[["target_end_date"]] - as.Date(reference_date)) / ts_temp_res),
+        .before = "horizon"
+      )
+  } else if (h_adjustments < 0) {
+    model_outputs <- extracted_outputs |>
+      dplyr::group_by(dplyr::across(c("transformation", "symmetrize", "window_size", "location", "output_type"))) |>
+      tidyr::complete(target_end_date = actual_target_dates, .data[["output_type_id"]]) |>
+      dplyr::ungroup() |>
```

Comment:
I don't understand this operation.

1. what is `complete` doing here? Maybe: adding in rows for extra `target_end_date`s that don't already show up in the `extracted_outputs`?  But why don't all of the needed `target_end_date`s show up already? This may be related to my comment/question on lines 72-81 above.
2. why is the `group_by` necessary?

Reply:
1. This is related to the discussion below for lines 135-138. Basically, we're adding in extra rows for target end dates whose predicted values are going to be replaced by observed data.
2. I found that `complete()` acts like `summarize()` in that it doesn't preserve columns not grouped across. Also, from the examples section "Within each 'group', [`complete()`] generates all possible combinations of [column values] that occur in the group"

Reply:
thanks for the explanation!